<a href="https://colab.research.google.com/github/Alyssasorensen/datasci_2_manipulation/blob/main/datasci_2_manipulation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Cleaning

For this week's homework, I will be data cleaning with Pandas, performing data transformation, and displaying alternative libraries.

### Loading Pandas Packages

In [88]:
import pandas as pd

### Loading Data

In [89]:
healthcare_data_cleaning = pd.read_csv('https://raw.githubusercontent.com/Alyssasorensen/datasci_2_manipulation/main/datasets/raw.githubusercontent.com_hantswilliams_HHA_507_2023_main_WK2_data_healthcare_data_cleaning.csv')

In [90]:
healthcare_data_cleaning

Patient Age  Gender  City of Residence State of Residence  \
0               45   Other         West Brian           Kentucky   
1               25  Female   East Jocelynfurt           Colorado   
2               51   Other  South Lindseyland               Ohio   
3               18  Female         Taylorfort       Pennsylvania   
4               48  Female          Birdmouth            Montana   
...            ...     ...                ...                ...   
104995          74   Other       West Anthony           Oklahoma   
104996          37    Male          Mckeetown               Utah   
104997          51  Female   Lake Lisachester           Nebraska   
104998     missing  Female       Heidiborough            missing   
104999          45  Female      South Jeffrey      Massachusetts   

       Has Insurance Visited Last Month Payment Method Preferred Doctor  \
0                Yes                Yes      Insurance     Dr. Williams   
1                 No                Yes           Card      Dr. Johnson   
2                Yes                 No           Cash     Dr. Williams   
3                Yes            missing           Cash     Dr. Williams   
4                 No                 No      Insurance        Dr. Brown   
...              ...                ...            ...              ...   
104995           Yes                 No           Card     Dr. Williams   
104996           Yes                 No           Cash        Dr. Jones   
104997           Yes                Yes      Insurance     Dr. Williams   
104998           Yes                 No           Cash        Dr. Smith   
104999            No                Yes           Cash        Dr. Brown   

       Disease Diagnosed Medication Prescribed Type of Appointment  \
0                    Flu                 Med_E             General   
1                    Flu                 Med_B          Specialist   
2               Covid-19                 Med_D             General   
3                    Flu                 Med_A             General   
4                   None                 Med_C          Specialist   
...                  ...                   ...                 ...   
104995          Covid-19                 Med_E          Specialist   
104996          Covid-19                 Med_C             General   
104997           Allergy                 Med_D           Follow-Up   
104998          Covid-19                 Med_A           Follow-Up   
104999           Allergy                 Med_D             General   

       Average Heart Rate Average BP Height (in cm) Weight (in kg)  \
0                      66        111            183             70   
1                      59    missing            174             58   
2                      79        119            161             56   
3                      99        115            171             52   
4                      93         89            186             87   
...                   ...        ...            ...            ...   
104995                 65        112            169             61   
104996            missing        111            165             67   
104997                 89        118            179             71   
104998                 74        123            167             75   
104999                 77        140            167             62   

           Payment Due ($) Last Visit (days ago) Visit Duration (mins)  \
0        65.45721578126224                   193                    92   
1       430.68365678679174                   195               missing   
2        315.0709305262176               missing                    37   
3        320.2998987723972                   175                   108   
4        264.2147372473319                   188                   113   
...                    ...                   ...                   ...   
104995  309.16444606680665                   183                    10   
104996         

### Changing "Missing" to "NaN"

In [91]:
healthcare_data_cleaning.replace('missing', pd.np.nan, inplace=True)
healthcare_data_cleaning

<ipython-input-91-724cc7559546>:1: FutureWarning: The pandas.np module is deprecated and will be removed from pandas in a future version. Import numpy directly instead.
  healthcare_data_cleaning.replace('missing', pd.np.nan, inplace=True)


Patient Age  Gender  City of Residence State of Residence  \
0               45   Other         West Brian           Kentucky   
1               25  Female   East Jocelynfurt           Colorado   
2               51   Other  South Lindseyland               Ohio   
3               18  Female         Taylorfort       Pennsylvania   
4               48  Female          Birdmouth            Montana   
...            ...     ...                ...                ...   
104995          74   Other       West Anthony           Oklahoma   
104996          37    Male          Mckeetown               Utah   
104997          51  Female   Lake Lisachester           Nebraska   
104998         NaN  Female       Heidiborough                NaN   
104999          45  Female      South Jeffrey      Massachusetts   

       Has Insurance Visited Last Month Payment Method Preferred Doctor  \
0                Yes                Yes      Insurance     Dr. Williams   
1                 No                Yes           Card      Dr. Johnson   
2                Yes                 No           Cash     Dr. Williams   
3                Yes                NaN           Cash     Dr. Williams   
4                 No                 No      Insurance        Dr. Brown   
...              ...                ...            ...              ...   
104995           Yes                 No           Card     Dr. Williams   
104996           Yes                 No           Cash        Dr. Jones   
104997           Yes                Yes      Insurance     Dr. Williams   
104998           Yes                 No           Cash        Dr. Smith   
104999            No                Yes           Cash        Dr. Brown   

       Disease Diagnosed Medication Prescribed Type of Appointment  \
0                    Flu                 Med_E             General   
1                    Flu                 Med_B          Specialist   
2               Covid-19                 Med_D             General   
3                    Flu                 Med_A             General   
4                   None                 Med_C          Specialist   
...                  ...                   ...                 ...   
104995          Covid-19                 Med_E          Specialist   
104996          Covid-19                 Med_C             General   
104997           Allergy                 Med_D           Follow-Up   
104998          Covid-19                 Med_A           Follow-Up   
104999           Allergy                 Med_D             General   

       Average Heart Rate Average BP Height (in cm) Weight (in kg)  \
0                      66        111            183             70   
1                      59        NaN            174             58   
2                      79        119            161             56   
3                      99        115            171             52   
4                      93         89            186             87   
...                   ...        ...            ...            ...   
104995                 65        112            169             61   
104996                NaN        111            165             67   
104997                 89        118            179             71   
104998                 74        123            167             75   
104999                 77        140            167             62   

           Payment Due ($) Last Visit (days ago) Visit Duration (mins)  \
0        65.45721578126224                   193                    92   
1       430.68365678679174                   195                   NaN   
2        315.0709305262176                   NaN                    37   
3        320.2998987723972                   175                   108   
4        264.2147372473319                   188                   113   
...                    ...                   ...                   ...   
104995  309.16444606680665                   183                    10   
104996         

### Removing Duplicate Rows

In [92]:
healthcare_data_cleaning.drop_duplicates(inplace=True)
healthcare_data_cleaning

Patient Age  Gender  City of Residence State of Residence  \
0               45   Other         West Brian           Kentucky   
1               25  Female   East Jocelynfurt           Colorado   
2               51   Other  South Lindseyland               Ohio   
3               18  Female         Taylorfort       Pennsylvania   
4               48  Female          Birdmouth            Montana   
...            ...     ...                ...                ...   
104994          53  Female      Michelleburgh       South Dakota   
104995          74   Other       West Anthony           Oklahoma   
104996          37    Male          Mckeetown               Utah   
104997          51  Female   Lake Lisachester           Nebraska   
104998         NaN  Female       Heidiborough                NaN   

       Has Insurance Visited Last Month Payment Method Preferred Doctor  \
0                Yes                Yes      Insurance     Dr. Williams   
1                 No                Yes           Card      Dr. Johnson   
2                Yes                 No           Cash     Dr. Williams   
3                Yes                NaN           Cash     Dr. Williams   
4                 No                 No      Insurance        Dr. Brown   
...              ...                ...            ...              ...   
104994            No                 No           Card     Dr. Williams   
104995           Yes                 No           Card     Dr. Williams   
104996           Yes                 No           Cash        Dr. Jones   
104997           Yes                Yes      Insurance     Dr. Williams   
104998           Yes                 No           Cash        Dr. Smith   

       Disease Diagnosed Medication Prescribed Type of Appointment  \
0                    Flu                 Med_E             General   
1                    Flu                 Med_B          Specialist   
2               Covid-19                 Med_D             General   
3                    Flu                 Med_A             General   
4                   None                 Med_C          Specialist   
...                  ...                   ...                 ...   
104994               Flu                 Med_B           Emergency   
104995          Covid-19                 Med_E          Specialist   
104996          Covid-19                 Med_C             General   
104997           Allergy                 Med_D           Follow-Up   
104998          Covid-19                 Med_A           Follow-Up   

       Average Heart Rate Average BP Height (in cm) Weight (in kg)  \
0                      66        111            183             70   
1                      59        NaN            174             58   
2                      79        119            161             56   
3                      99        115            171             52   
4                      93         89            186             87   
...                   ...        ...            ...            ...   
104994                 68        109            172             75   
104995                 65        112            169             61   
104996                NaN        111            165             67   
104997                 89        118            179             71   
104998                 74        123            167             75   

           Payment Due ($) Last Visit (days ago) Visit Duration (mins)  \
0        65.45721578126224                   193                    92   
1       430.68365678679174                   195                   NaN   
2        315.0709305262176                   NaN                    37   
3        320.2998987723972                   175                   108   
4        264.2147372473319                   188                   113   
...                    ...                   ...                   ...   
104994   488.0316041186948                   167                    25   
104995  309.164

### Removing White Space and Special Characters

In [93]:
import re  # Import the 're' module

# Function to remove white space and special characters from a value
def clean_column_names(healthcare_data_cleaning):
    # Define a helper function to clean column names
    def clean_name(name):
        cleaned_name = re.sub(r'[^a-zA-Z0-9]', '', name)
        return cleaned_name.lower()

    # Rename columns using the helper function
    # This is using a list comprehension - e.g., we have a list to the right of the equals sign,
    # and inside the list, we are applying our function for every col (or X) that exists in df.columns
    healthcare_data_cleaning.columns = [clean_name(col) for col in healthcare_data_cleaning.columns]  # Use 'df' instead of 'healthcare_data_cleaning'
    return healthcare_data_cleaning

# Apply the clean_column_names function to all columns
healthcare_data_cleaning = clean_column_names(healthcare_data_cleaning)

print(healthcare_data_cleaning)


       patientage  gender    cityofresidence stateofresidence hasinsurance  \
0              45   Other         West Brian         Kentucky          Yes   
1              25  Female   East Jocelynfurt         Colorado           No   
2              51   Other  South Lindseyland             Ohio          Yes   
3              18  Female         Taylorfort     Pennsylvania          Yes   
4              48  Female          Birdmouth          Montana           No   
...           ...     ...                ...              ...          ...   
104994         53  Female      Michelleburgh     South Dakota           No   
104995         74   Other       West Anthony         Oklahoma          Yes   
104996         37    Male          Mckeetown             Utah          Yes   
104997         51  Female   Lake Lisachester         Nebraska          Yes   
104998        NaN  Female       Heidiborough              NaN          Yes   

       visitedlastmonth paymentmethod preferreddoctor diseasedi

### Handling Missing Values

In [94]:
import re
import numpy as np

In [95]:
healthcare_data_cleaning['paymentdue'] = pd.to_numeric(healthcare_data_cleaning['paymentdue'], errors='coerce')
healthcare_data_cleaning['prescriptioncost'] = pd.to_numeric(healthcare_data_cleaning['prescriptioncost'], errors='coerce')

In [96]:
healthcare_data_cleaning['patientage'] = pd.to_numeric(healthcare_data_cleaning['patientage'], errors='coerce')
healthcare_data_cleaning['averageheartrate'] = pd.to_numeric(healthcare_data_cleaning['averageheartrate'], errors='coerce')
healthcare_data_cleaning['averagebp'] = pd.to_numeric(healthcare_data_cleaning['averagebp'], errors='coerce')
healthcare_data_cleaning['heightincm'] = pd.to_numeric(healthcare_data_cleaning['heightincm'], errors='coerce')
healthcare_data_cleaning['weightinkg'] = pd.to_numeric(healthcare_data_cleaning['weightinkg'], errors='coerce')

In [97]:
healthcare_data_cleaning['lastvisitdaysago'] = pd.to_numeric(healthcare_data_cleaning['lastvisitdaysago'], errors='coerce')
healthcare_data_cleaning['visitdurationmins'] = pd.to_numeric(healthcare_data_cleaning['visitdurationmins'], errors='coerce')
healthcare_data_cleaning['numberoftests'] = pd.to_numeric(healthcare_data_cleaning['numberoftests'], errors='coerce')

In [98]:
print(healthcare_data_cleaning['patientage'].value_counts())

18.0     4047
47.0     2555
45.0     2543
44.0     2529
42.0     2518
         ... 
100.0      10
96.0       10
97.0        8
99.0        7
98.0        5
Name: patientage, Length: 83, dtype: int64


In [99]:
print(healthcare_data_cleaning['gender'].value_counts())

Male      31728
Female    31674
Other     31598
Name: gender, dtype: int64


In [100]:
print(healthcare_data_cleaning['cityofresidence'].value_counts())

Lake Michael         100
West Michael          82
East Michael          77
Port Michael          74
North Michael         71
                    ... 
Deleontown             1
Lake Hollychester      1
Port Edwardhaven       1
West Pamelaton         1
Mckeetown              1
Name: cityofresidence, Length: 37132, dtype: int64


In [101]:
print(healthcare_data_cleaning['stateofresidence'].value_counts())

New Mexico        1969
Nevada            1969
Idaho             1965
Massachusetts     1959
California        1952
Ohio              1950
Connecticut       1950
Kentucky          1946
North Carolina    1944
New Jersey        1932
Alaska            1927
Wisconsin         1926
Nebraska          1922
Rhode Island      1918
Minnesota         1918
Wyoming           1917
West Virginia     1916
Arizona           1915
Tennessee         1914
South Carolina    1912
Virginia          1911
North Dakota      1911
Utah              1910
Kansas            1909
Pennsylvania      1908
Missouri          1906
Iowa              1906
Montana           1904
Mississippi       1897
Florida           1893
Louisiana         1893
New York          1891
Illinois          1891
Oregon            1888
Indiana           1886
Vermont           1885
Oklahoma          1876
Texas             1876
Alabama           1875
Michigan          1863
Georgia           1862
Colorado          1859
Delaware          1859
Hawaii     

In [18]:
print(healthcare_data_cleaning['hasinsurance'].value_counts())

Yes    47597
No     47403
Name: hasinsurance, dtype: int64


In [19]:
print(healthcare_data_cleaning['visitedlastmonth'].value_counts())

No     47572
Yes    47428
Name: visitedlastmonth, dtype: int64


In [20]:
print(healthcare_data_cleaning['paymentmethod'].value_counts())

Card         31744
Insurance    31662
Cash         31594
Name: paymentmethod, dtype: int64


In [21]:
print(healthcare_data_cleaning['preferreddoctor'].value_counts())

Dr. Jones       19153
Dr. Williams    19078
Dr. Johnson     19005
Dr. Smith       18914
Dr. Brown       18850
Name: preferreddoctor, dtype: int64


In [22]:
print(healthcare_data_cleaning['diseasediagnosed'].value_counts())

Flu         19123
Allergy     19063
None        19021
Cold        18962
Covid-19    18831
Name: diseasediagnosed, dtype: int64


In [23]:
print(healthcare_data_cleaning['medicationprescribed'].value_counts())

Med_C    19149
Med_D    19040
Med_B    19003
Med_A    18957
Med_E    18851
Name: medicationprescribed, dtype: int64


In [24]:
print(healthcare_data_cleaning['typeofappointment'].value_counts())

Specialist    23841
Follow-Up     23832
Emergency     23737
General       23590
Name: typeofappointment, dtype: int64


In [25]:
print(healthcare_data_cleaning['averageheartrate'].value_counts())

80     3779
81     3769
78     3764
77     3711
82     3694
       ... 
114      11
116       6
120       4
117       3
119       1
Name: averageheartrate, Length: 70, dtype: int64


In [26]:
print(healthcare_data_cleaning['averagebp'].value_counts())

80     2560
110    2509
109    2509
112    2491
111    2489
       ... 
82      489
137     465
81      425
138     415
139     374
Name: averagebp, Length: 61, dtype: int64


In [27]:
print(healthcare_data_cleaning['heightincm'].value_counts())

174    3834
175    3767
176    3731
177    3727
173    3702
       ... 
142      19
207      17
141      15
208      12
209      11
Name: heightincm, Length: 71, dtype: int64


In [28]:
print(healthcare_data_cleaning['weightinkg'].value_counts())

71     2559
40     2544
67     2532
72     2526
69     2481
       ... 
134       1
130       1
132       1
128       1
135       1
Name: weightinkg, Length: 93, dtype: int64


In [29]:
print(healthcare_data_cleaning['paymentdue'].value_counts())

65.45721578126224     1
299.98849951773184    1
150.18465367858602    1
297.9640902267887     1
29.961924251165616    1
                     ..
471.8302625604913     1
456.18615965963943    1
273.77175625186544    1
315.28904224377254    1
5.341231420036174     1
Name: paymentdue, Length: 95000, dtype: int64


In [30]:
print(healthcare_data_cleaning['lastvisitdaysago'].value_counts())

183    3992
182    3985
184    3927
180    3789
185    3777
       ... 
223       2
140       1
220       1
221       1
222       1
Name: lastvisitdaysago, Length: 81, dtype: int64


In [31]:
print(healthcare_data_cleaning['visitdurationmins'].value_counts())

49     931
69     924
68     917
27     914
36     906
      ... 
98     816
109    809
87     808
67     804
107    802
Name: visitdurationmins, Length: 111, dtype: int64


In [32]:
print(healthcare_data_cleaning['numberoftests'].value_counts())

4     23795
3     20400
5     19094
2     11541
6     10566
7      3993
1      3925
8       979
0       537
9       160
10       10
Name: numberoftests, dtype: int64


In [33]:
print(healthcare_data_cleaning['prescriptioncost'].value_counts())

15.71751735532889     1
65.99296531902111     1
134.85104405526357    1
67.13960005163149     1
4.058180013509604     1
                     ..
86.08310504344813     1
50.48852584305967     1
78.53336852897506     1
124.01984170637614    1
124.4083409893742     1
Name: prescriptioncost, Length: 95000, dtype: int64


In [34]:
missing_values = healthcare_data_cleaning.isnull().sum()
missing_values

patientage              5000
gender                  5000
cityofresidence         5000
stateofresidence        5000
hasinsurance            5000
visitedlastmonth        5000
paymentmethod           5000
preferreddoctor         5000
diseasediagnosed        5000
medicationprescribed    5000
typeofappointment       5000
averageheartrate        5000
averagebp               5000
heightincm              5000
weightinkg              5000
paymentdue              5000
lastvisitdaysago        5000
visitdurationmins       5000
numberoftests           5000
prescriptioncost        5000
dtype: int64

In [102]:
healthcare_data_cleaning = healthcare_data_cleaning.replace('missing', 'Unknown')
healthcare_data_cleaning

patientage  gender    cityofresidence stateofresidence hasinsurance  \
0             45.0   Other         West Brian         Kentucky          Yes   
1             25.0  Female   East Jocelynfurt         Colorado           No   
2             51.0   Other  South Lindseyland             Ohio          Yes   
3             18.0  Female         Taylorfort     Pennsylvania          Yes   
4             48.0  Female          Birdmouth          Montana           No   
...            ...     ...                ...              ...          ...   
104994        53.0  Female      Michelleburgh     South Dakota           No   
104995        74.0   Other       West Anthony         Oklahoma          Yes   
104996        37.0    Male          Mckeetown             Utah          Yes   
104997        51.0  Female   Lake Lisachester         Nebraska          Yes   
104998         NaN  Female       Heidiborough              NaN          Yes   

       visitedlastmonth paymentmethod preferreddoctor diseasediagnosed  \
0                   Yes     Insurance    Dr. Williams              Flu   
1                   Yes          Card     Dr. Johnson              Flu   
2                    No          Cash    Dr. Williams         Covid-19   
3                   NaN          Cash    Dr. Williams              Flu   
4                    No     Insurance       Dr. Brown             None   
...                 ...           ...             ...              ...   
104994               No          Card    Dr. Williams              Flu   
104995               No          Card    Dr. Williams         Covid-19   
104996               No          Cash       Dr. Jones         Covid-19   
104997              Yes     Insurance    Dr. Williams          Allergy   
104998               No          Cash       Dr. Smith         Covid-19   

       medicationprescribed typeofappointment  averageheartrate  averagebp  \
0                     Med_E           General              66.0      111.0   
1                     Med_B        Specialist              59.0        NaN   
2                     Med_D           General              79.0      119.0   
3                     Med_A           General              99.0      115.0   
4                     Med_C        Specialist              93.0       89.0   
...                     ...               ...               ...        ...   
104994                Med_B         Emergency              68.0      109.0   
104995                Med_E        Specialist              65.0      112.0   
104996                Med_C           General               NaN      111.0   
104997                Med_D         Follow-Up              89.0      118.0   
104998                Med_A         Follow-Up              74.0      123.0   

        heightincm  weightinkg  paymentdue  lastvisitdaysago  \
0            183.0        70.0   65.457216             193.0   
1            174.0        58.0  430.683657             195.0   
2            161.0        56.0  315.070931               NaN   
3            171.0        52.0  320.299899             175.0   
4            186.0        87.0  264.214737             188.0   
...            ...         ...         ...               ...   
104994       172.0        75.0  488.031604             167.0   
104995       169.0        61.0  309.164446             183.0   
104996       165.0        67.0         NaN             191.0   
104997       179.0        71.0  320.830418             172.0   
104998       167.0        75.0    5.341231             180.0   

        visitdurationmins  numberoftests  prescriptioncost  
0                    92.0            5.0         15.717517  
1                     NaN            4.0         80.786473  
2                    37.0            2.0         64.392139  
3                   108.0            2.0          8.872859  
4                   113.0            3.0         77.483113  
...                   ...            ...               ...  
104994               25.0            

### Cleaning Column Names

In [103]:
healthcare_data_cleaning.columns = healthcare_data_cleaning.columns.str.strip().str.lower().str.replace(' ', '_').str.replace('(', '').str.replace(')', '')
healthcare_data_cleaning

<ipython-input-103-bde3266830c7>:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  healthcare_data_cleaning.columns = healthcare_data_cleaning.columns.str.strip().str.lower().str.replace(' ', '_').str.replace('(', '').str.replace(')', '')
<ipython-input-103-bde3266830c7>:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  healthcare_data_cleaning.columns = healthcare_data_cleaning.columns.str.strip().str.lower().str.replace(' ', '_').str.replace('(', '').str.replace(')', '')


patientage  gender    cityofresidence stateofresidence hasinsurance  \
0             45.0   Other         West Brian         Kentucky          Yes   
1             25.0  Female   East Jocelynfurt         Colorado           No   
2             51.0   Other  South Lindseyland             Ohio          Yes   
3             18.0  Female         Taylorfort     Pennsylvania          Yes   
4             48.0  Female          Birdmouth          Montana           No   
...            ...     ...                ...              ...          ...   
104994        53.0  Female      Michelleburgh     South Dakota           No   
104995        74.0   Other       West Anthony         Oklahoma          Yes   
104996        37.0    Male          Mckeetown             Utah          Yes   
104997        51.0  Female   Lake Lisachester         Nebraska          Yes   
104998         NaN  Female       Heidiborough              NaN          Yes   

       visitedlastmonth paymentmethod preferreddoctor diseasediagnosed  \
0                   Yes     Insurance    Dr. Williams              Flu   
1                   Yes          Card     Dr. Johnson              Flu   
2                    No          Cash    Dr. Williams         Covid-19   
3                   NaN          Cash    Dr. Williams              Flu   
4                    No     Insurance       Dr. Brown             None   
...                 ...           ...             ...              ...   
104994               No          Card    Dr. Williams              Flu   
104995               No          Card    Dr. Williams         Covid-19   
104996               No          Cash       Dr. Jones         Covid-19   
104997              Yes     Insurance    Dr. Williams          Allergy   
104998               No          Cash       Dr. Smith         Covid-19   

       medicationprescribed typeofappointment  averageheartrate  averagebp  \
0                     Med_E           General              66.0      111.0   
1                     Med_B        Specialist              59.0        NaN   
2                     Med_D           General              79.0      119.0   
3                     Med_A           General              99.0      115.0   
4                     Med_C        Specialist              93.0       89.0   
...                     ...               ...               ...        ...   
104994                Med_B         Emergency              68.0      109.0   
104995                Med_E        Specialist              65.0      112.0   
104996                Med_C           General               NaN      111.0   
104997                Med_D         Follow-Up              89.0      118.0   
104998                Med_A         Follow-Up              74.0      123.0   

        heightincm  weightinkg  paymentdue  lastvisitdaysago  \
0            183.0        70.0   65.457216             193.0   
1            174.0        58.0  430.683657             195.0   
2            161.0        56.0  315.070931               NaN   
3            171.0        52.0  320.299899             175.0   
4            186.0        87.0  264.214737             188.0   
...            ...         ...         ...               ...   
104994       172.0        75.0  488.031604             167.0   
104995       169.0        61.0  309.164446             183.0   
104996       165.0        67.0         NaN             191.0   
104997       179.0        71.0  320.830418             172.0   
104998       167.0        75.0    5.341231             180.0   

        visitdurationmins  numberoftests  prescriptioncost  
0                    92.0            5.0         15.717517  
1                     NaN            4.0         80.786473  
2                    37.0            2.0         64.392139  
3                   108.0            2.0          8.872859  
4                   113.0            3.0         77.483113  
...                   ...            ...               ...  
104994               25.0            

### Summary of the Changes Made

##### The initial step involved loading the packages into Google Colab. This meant that I needed to write "import pandas as pd." Following this, I had to load the data. I inserted the CSV file from GitHub and Google Colab displayed the data. Following this I used, "healthcare_data_cleaning.replace('missing', pd.np.nan, inplace=True)" to replace the "missing" values with "NaN". After this, I removed duplicate rows from the DataFrame. This was done by inserting, "healthcare_data_cleaning.drop_duplicates(inplace=True)." The "drop_duplicates" identifies and removes duplicate rows from the DataFrame based on the columns' values by default. The statement "inplace=True" is an argument passed to the "drop_duplicates" method. When "inplace=True" is set, it means the operation wil be performed directly on the original DataFrame, which is healthcare_data_cleaning. After this, I removed white spaces and special characters. The function inserted for this was used to ensure that column names in the DataFrame do not contain non-alphanumeric characters, white spaces, and are standardized to lowercase, which can be helpful for consistency. Next, I added the decimals back to the values of payment due and prescription cost columns. Next, I handled the missing values from the DataFrame. I did this by inserting "print(data_types['patientage'].value_counts())." Following this, I cleaned some of the column names by changing how they are displayed in the DataFrame.

# Data Transformation




### Creating New Columns Based on Existing Ones

In [104]:
healthcare_data_cleaning.rename(columns = {'patient_age':'age', 'city_of_residence': 'city', 'state_of_residence': 'state', 'disease_diagnosed':'diagnosis', 'medication_prescribed': 'prescription', 'height_in_cm': 'height_cm', 'weight_in_kg':'weight_kg'}, inplace = True)
healthcare_data_cleaning

patientage  gender    cityofresidence stateofresidence hasinsurance  \
0             45.0   Other         West Brian         Kentucky          Yes   
1             25.0  Female   East Jocelynfurt         Colorado           No   
2             51.0   Other  South Lindseyland             Ohio          Yes   
3             18.0  Female         Taylorfort     Pennsylvania          Yes   
4             48.0  Female          Birdmouth          Montana           No   
...            ...     ...                ...              ...          ...   
104994        53.0  Female      Michelleburgh     South Dakota           No   
104995        74.0   Other       West Anthony         Oklahoma          Yes   
104996        37.0    Male          Mckeetown             Utah          Yes   
104997        51.0  Female   Lake Lisachester         Nebraska          Yes   
104998         NaN  Female       Heidiborough              NaN          Yes   

       visitedlastmonth paymentmethod preferreddoctor diseasediagnosed  \
0                   Yes     Insurance    Dr. Williams              Flu   
1                   Yes          Card     Dr. Johnson              Flu   
2                    No          Cash    Dr. Williams         Covid-19   
3                   NaN          Cash    Dr. Williams              Flu   
4                    No     Insurance       Dr. Brown             None   
...                 ...           ...             ...              ...   
104994               No          Card    Dr. Williams              Flu   
104995               No          Card    Dr. Williams         Covid-19   
104996               No          Cash       Dr. Jones         Covid-19   
104997              Yes     Insurance    Dr. Williams          Allergy   
104998               No          Cash       Dr. Smith         Covid-19   

       medicationprescribed typeofappointment  averageheartrate  averagebp  \
0                     Med_E           General              66.0      111.0   
1                     Med_B        Specialist              59.0        NaN   
2                     Med_D           General              79.0      119.0   
3                     Med_A           General              99.0      115.0   
4                     Med_C        Specialist              93.0       89.0   
...                     ...               ...               ...        ...   
104994                Med_B         Emergency              68.0      109.0   
104995                Med_E        Specialist              65.0      112.0   
104996                Med_C           General               NaN      111.0   
104997                Med_D         Follow-Up              89.0      118.0   
104998                Med_A         Follow-Up              74.0      123.0   

        heightincm  weightinkg  paymentdue  lastvisitdaysago  \
0            183.0        70.0   65.457216             193.0   
1            174.0        58.0  430.683657             195.0   
2            161.0        56.0  315.070931               NaN   
3            171.0        52.0  320.299899             175.0   
4            186.0        87.0  264.214737             188.0   
...            ...         ...         ...               ...   
104994       172.0        75.0  488.031604             167.0   
104995       169.0        61.0  309.164446             183.0   
104996       165.0        67.0         NaN             191.0   
104997       179.0        71.0  320.830418             172.0   
104998       167.0        75.0    5.341231             180.0   

        visitdurationmins  numberoftests  prescriptioncost  
0                    92.0            5.0         15.717517  
1                     NaN            4.0         80.786473  
2                    37.0            2.0         64.392139  
3                   108.0            2.0          8.872859  
4                   113.0            3.0         77.483113  
...                   ...            ...               ...  
104994               25.0            

### Aggregating Data Using Groupby and Compute Summary Statistics

In [105]:
print(healthcare_data_cleaning['paymentdue'].dtypes)

float64


In [106]:
healthcare_data_cleaning['paymentdue'] = pd.to_numeric(healthcare_data_cleaning['paymentdue'], errors='coerce')

In [107]:
healthcare_data_cleaning['diseasediagnosed'].value_counts()

Flu         19123
Allergy     19063
None        19021
Cold        18962
Covid-19    18831
Name: diseasediagnosed, dtype: int64

In [108]:
grouped_data = healthcare_data_cleaning.groupby('diseasediagnosed').agg({
    'paymentdue': 'mean'
})
grouped_data

paymentdue
diseasediagnosed            
Allergy           249.602752
Cold              249.469804
Covid-19          250.307367
Flu               248.349807
None              250.790789

In [109]:
healthcare_data_cleaning

patientage  gender    cityofresidence stateofresidence hasinsurance  \
0             45.0   Other         West Brian         Kentucky          Yes   
1             25.0  Female   East Jocelynfurt         Colorado           No   
2             51.0   Other  South Lindseyland             Ohio          Yes   
3             18.0  Female         Taylorfort     Pennsylvania          Yes   
4             48.0  Female          Birdmouth          Montana           No   
...            ...     ...                ...              ...          ...   
104994        53.0  Female      Michelleburgh     South Dakota           No   
104995        74.0   Other       West Anthony         Oklahoma          Yes   
104996        37.0    Male          Mckeetown             Utah          Yes   
104997        51.0  Female   Lake Lisachester         Nebraska          Yes   
104998         NaN  Female       Heidiborough              NaN          Yes   

       visitedlastmonth paymentmethod preferreddoctor diseasediagnosed  \
0                   Yes     Insurance    Dr. Williams              Flu   
1                   Yes          Card     Dr. Johnson              Flu   
2                    No          Cash    Dr. Williams         Covid-19   
3                   NaN          Cash    Dr. Williams              Flu   
4                    No     Insurance       Dr. Brown             None   
...                 ...           ...             ...              ...   
104994               No          Card    Dr. Williams              Flu   
104995               No          Card    Dr. Williams         Covid-19   
104996               No          Cash       Dr. Jones         Covid-19   
104997              Yes     Insurance    Dr. Williams          Allergy   
104998               No          Cash       Dr. Smith         Covid-19   

       medicationprescribed typeofappointment  averageheartrate  averagebp  \
0                     Med_E           General              66.0      111.0   
1                     Med_B        Specialist              59.0        NaN   
2                     Med_D           General              79.0      119.0   
3                     Med_A           General              99.0      115.0   
4                     Med_C        Specialist              93.0       89.0   
...                     ...               ...               ...        ...   
104994                Med_B         Emergency              68.0      109.0   
104995                Med_E        Specialist              65.0      112.0   
104996                Med_C           General               NaN      111.0   
104997                Med_D         Follow-Up              89.0      118.0   
104998                Med_A         Follow-Up              74.0      123.0   

        heightincm  weightinkg  paymentdue  lastvisitdaysago  \
0            183.0        70.0   65.457216             193.0   
1            174.0        58.0  430.683657             195.0   
2            161.0        56.0  315.070931               NaN   
3            171.0        52.0  320.299899             175.0   
4            186.0        87.0  264.214737             188.0   
...            ...         ...         ...               ...   
104994       172.0        75.0  488.031604             167.0   
104995       169.0        61.0  309.164446             183.0   
104996       165.0        67.0         NaN             191.0   
104997       179.0        71.0  320.830418             172.0   
104998       167.0        75.0    5.341231             180.0   

        visitdurationmins  numberoftests  prescriptioncost  
0                    92.0            5.0         15.717517  
1                     NaN            4.0         80.786473  
2                    37.0            2.0         64.392139  
3                   108.0            2.0          8.872859  
4                   113.0            3.0         77.483113  
...                   ...            ...               ...  
104994               25.0            

In [110]:
encompass_diagnosis = healthcare_data_cleaning[healthcare_data_cleaning['diseasediagnosed'] == 'Covid-19']
encompass_diagnosis.groupby(['diseasediagnosed']).sum()

<ipython-input-110-04a36134b7ba>:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  encompass_diagnosis.groupby(['diseasediagnosed']).sum()


patientage  averageheartrate  averagebp  heightincm  \
diseasediagnosed                                                        
Covid-19            796229.0         1423634.0  1954912.0   3133540.0   

                  weightinkg    paymentdue  lastvisitdaysago  \
diseasediagnosed                                               
Covid-19           1241259.0  4.478499e+06         3272427.0   

                  visitdurationmins  numberoftests  prescriptioncost  
diseasediagnosed                                                      
Covid-19                  1162782.0        71489.0      1.343962e+06

In [111]:
healthcare_data_cleaning.columns

Index(['patientage', 'gender', 'cityofresidence', 'stateofresidence',
       'hasinsurance', 'visitedlastmonth', 'paymentmethod', 'preferreddoctor',
       'diseasediagnosed', 'medicationprescribed', 'typeofappointment',
       'averageheartrate', 'averagebp', 'heightincm', 'weightinkg',
       'paymentdue', 'lastvisitdaysago', 'visitdurationmins', 'numberoftests',
       'prescriptioncost'],
      dtype='object')

In [112]:
pivot_data1 = healthcare_data_cleaning.pivot_table(index='cityofresidence', columns='diseasediagnosed', values='paymentdue', aggfunc='mean')
pivot_data1

diseasediagnosed     Allergy        Cold    Covid-19         Flu        None
cityofresidence                                                             
Aaronberg          53.435276         NaN         NaN  200.185919         NaN
Aaronborough      174.035005  425.948239  417.036905         NaN  101.336997
Aaronburgh        302.079018  163.291124  191.207121  426.693173  348.700014
Aaronbury                NaN         NaN  217.983568  183.814863         NaN
Aaronchester      331.706527         NaN  144.269185         NaN  290.556463
...                      ...         ...         ...         ...         ...
Zoeburgh                 NaN         NaN         NaN  458.810402         NaN
Zoechester               NaN  142.003169         NaN         NaN         NaN
Zunigaland               NaN         NaN         NaN  346.871637         NaN
Zunigamouth              NaN         NaN         NaN  379.164892         NaN
Zunigaport               NaN  265.356088         NaN         NaN         NaN

[35083 rows x 5 columns]

In [113]:
pivot_table = healthcare_data_cleaning.pivot_table(index='gender', columns='hasinsurance', values='paymentdue', aggfunc='mean')
pivot_table

hasinsurance          No         Yes
gender                              
Female        248.505546  249.048480
Male          251.145420  249.736330
Other         248.469690  250.898797

In [114]:
pivot_data3 = healthcare_data_cleaning.pivot_table(index='preferreddoctor', columns='typeofappointment', values='numberoftests', aggfunc='median')
pivot_data3

typeofappointment  Emergency  Follow-Up  General  Specialist
preferreddoctor                                             
Dr. Brown                4.0        4.0      4.0         4.0
Dr. Johnson              4.0        4.0      4.0         4.0
Dr. Jones                4.0        4.0      4.0         4.0
Dr. Smith                4.0        4.0      4.0         4.0
Dr. Williams             4.0        4.0      4.0         4.0

# Introduction to Alternative Libraries

### Loading Polars Package

In [115]:
import polars as pl

### Loading Data for Polars

In [116]:
healthcare_data_cleaning = pl.read_csv('https://raw.githubusercontent.com/Alyssasorensen/datasci_2_manipulation/main/datasets/raw.githubusercontent.com_hantswilliams_HHA_507_2023_main_WK2_data_healthcare_data_cleaning.csv')

In [117]:
healthcare_data_cleaning

shape: (105_000, 20)
┌───────────┬────────┬────────────┬────────────┬───┬────────────┬────────────┬────────────┬────────────┐
│ Patient   ┆ Gender ┆ City of    ┆ State of   ┆ … ┆ Last Visit ┆ Visit      ┆ Number of  ┆ Prescripti │
│ Age       ┆ ---    ┆ Residence  ┆ Residence  ┆   ┆ (days ago) ┆ Duration   ┆ Tests      ┆ on Cost    │
│ ---       ┆ str    ┆ ---        ┆ ---        ┆   ┆ ---        ┆ (mins)     ┆ ---        ┆ ($)        │
│ str       ┆        ┆ str        ┆ str        ┆   ┆ str        ┆ ---        ┆ str        ┆ ---        │
│           ┆        ┆            ┆            ┆   ┆            ┆ str        ┆            ┆ str        │
╞═══════════╪════════╪════════════╪════════════╪═══╪════════════╪════════════╪════════════╪════════════╡
│ 45        ┆ Other  ┆ West Brian ┆ Kentucky   ┆ … ┆ 193        ┆ 92         ┆ 5          ┆ 15.7175173 │
│           ┆        ┆            ┆            ┆   ┆            ┆            ┆            ┆ 5532889    │
│ 25        ┆ Female ┆ East Jocel ┆ Colorado   ┆ … ┆ 195        ┆ missing    ┆ 4          ┆ 80.7864728 │
│           ┆        ┆ ynfurt     ┆            ┆   ┆            ┆            ┆            ┆ 4463952    │
│ 51        ┆ Other  ┆ South Lind ┆ Ohio       ┆ … ┆ missing    ┆ 37         ┆ 2          ┆ 64.3921393 │
│           ┆        ┆ seyland    ┆            ┆   ┆            ┆            ┆            ┆ 009105     │
│ 18        ┆ Female ┆ Taylorfort ┆ Pennsylvan ┆ … ┆ 175        ┆ 108        ┆ 2          ┆ 8.87285921 │
│           ┆        ┆            ┆ ia         ┆   ┆            ┆            ┆            ┆ 5315316    │
│ …         ┆ …      ┆ …          ┆ …          ┆ … ┆ …          ┆ …          ┆ …          ┆ …          │
│ 37        ┆ Male   ┆ Mckeetown  ┆ Utah       ┆ … ┆ 191        ┆ 47         ┆ 2          ┆ 113.867640 │
│           ┆        ┆            ┆            ┆   ┆            ┆            ┆            ┆ 25370326   │
│ 51        ┆ Female ┆ Lake Lisac ┆ Nebraska   ┆ … ┆ 172        ┆ 39         ┆ 3          ┆ 35.6224202 │
│           ┆        ┆ hester     ┆            ┆   ┆            ┆            ┆            ┆ 32475356   │
│ missing   ┆ Female ┆ Heidiborou ┆ missing    ┆ … ┆ 180        ┆ 79         ┆ missing    ┆ 124.408340 │
│           ┆        ┆ gh         ┆            ┆   ┆            ┆            ┆            ┆ 9893742    │
│ 45        ┆ Female ┆ South      ┆ Massachuse ┆ … ┆ 167        ┆ 81         ┆ missing    ┆ 87.9106014 │
│           ┆        ┆ Jeffrey    ┆ tts        ┆   ┆            ┆            ┆            ┆ 9353706    │
└───────────┴────────┴────────────┴────────────┴───┴────────────┴────────────┴────────────┴────────────┘

In [118]:
import polars as pl

healthcare_data_cleaning = pl.read_csv('https://raw.githubusercontent.com/Alyssasorensen/datasci_2_manipulation/main/datasets/raw.githubusercontent.com_hantswilliams_HHA_507_2023_main_WK2_data_healthcare_data_cleaning.csv')

print("Number of rows:", healthcare_data_cleaning.shape[0])
print("Number of columns:", healthcare_data_cleaning.shape[1])

Number of rows: 105000
Number of columns: 20


In [119]:
import time

In [120]:
dataset_path = 'healthcare_data_cleaning.csv'

In [121]:
start_time = time.time()
polars_healthcare_data_cleaning = pd.read_csv('https://raw.githubusercontent.com/Alyssasorensen/datasci_2_manipulation/main/datasets/raw.githubusercontent.com_hantswilliams_HHA_507_2023_main_WK2_data_healthcare_data_cleaning.csv')
polars_filtered = polars_healthcare_data_cleaning[polars_healthcare_data_cleaning['Gender'] == 'Male']
polars_average_age = polars_filtered['Average BP'].sum()
polars_time = time.time() - start_time

### Loading Times

In [122]:
print(f"Polars Execution Time: {polars_time:.4f} seconds")

Polars Execution Time: 0.6333 seconds


### Loading Times Observation
###### When I first began the assignment I used both Modin and Polars. Modin took longer than Polars to execute the code. Modin took 0.5426 seconds and Polars took 0.4323 seconds. Modin took approximately 0.1103 seconds longer to execute the code. However, I had to remove Modin from my code because it would not work with the code that I used to change the "missing" values into the "NaN" values. When I did the load time for Polars a second time it took 0.6333 seconds.  